### PR TITLE
fix: remove subscription when correlation is rejected

### LINKED ISSUE #14814: Leaked message subscriptions cause constant processing load
On a Zeebe 8.2.12 Cluster we observed the following load. This does not resolve itself and the load remains constant.

```
COMMAND PROCESS_MESSAGE_SUBSCRIPTION CORRELATE 22517998138010331 
COMMAND PROCESS_MESSAGE_SUBSCRIPTION CORRELATE 22517998139003162 
COMMAND PROCESS_MESSAGE_SUBSCRIPTION CORRELATE 22517998140261067 
COMMAND PROCESS_MESSAGE_SUBSCRIPTION CORRELATE 22517998142789239 
...
COMMAND_REJECTION PROCESS_MESSAGE_SUBSCRIPTION CORRELATE 22517998138010331 
COMMAND MESSAGE_SUBSCRIPTION REJECT 22517998138010331 
COMMAND_REJECTION MESSAGE_SUBSCRIPTION REJECT 22517998138010331 
COMMAND_REJECTION PROCESS_MESSAGE_SUBSCRIPTION CORRELATE 22517998139003162 
COMMAND MESSAGE_SUBSCRIPTION REJECT 22517998139003162 
COMMAND_REJECTION MESSAGE_SUBSCRIPTION REJECT 22517998139003162 
COMMAND_REJECTION PROCESS_MESSAGE_SUBSCRIPTION CORRELATE 22517998140261067 
COMMAND MESSAGE_SUBSCRIPTION REJECT 22517998140261067 
COMMAND_REJECTION MESSAGE_SUBSCRIPTION REJECT 22517998140261067 
COMMAND_REJECTION PROCESS_MESSAGE_SUBSCRIPTION CORRELATE 22517998142789239 
COMMAND MESSAGE_SUBSCRIPTION REJECT 22517998142789239 
COMMAND_REJECTION MESSAGE_SUBSCRIPTION REJECT 22517998142789239 
```

The `PendingMessageSubscriptionChecker` writes `PROCESS_MESSAGE_SUBSCRIPTION/CORRELATE` commands for entries in `MutablePendingMessageSubscriptionState`. While processing these commands, the `ProcessMessageSubscriptionCorrelateProcessor` doesn't find a matching subscription and writes a `COMMAND_REJECTION` and a `MESSAGE_SUBSCRIPTION/REJECT` command. The `MessageSubscriptionRejectProcessor` _also_ rejects the `MESSAGE_SUBSCRIPTION/REJECT` command by writing a `COMMAND_REJECTION` because it does not find a matching correlation in the message state.

The `PROCESS_MESSAGE_SUBSCRIPTION/CORRELATE` is rejected with cause `NOT_FOUND` and a matching message: "Expected to correlate process message subscription with element key '22517998206768899' and message name 'MY_MESSAGE_NAME', but no such subscription was found".
The `MESSAGE_SUBSCRIPTION/REJECT` is rejected with cause `INVALID_STATE` and message "Expected message '112589990754201722' to be correlated for process with BPMN process id 'MY_PROCESS_ID' but no correlation was found"

## Cause

On one partition we have the subscription column family containing a subscription with `correlating=true` so the `PROCESS_MESSAGE_SUBSCRIPTION/CORRELATE` is sent again and again. On the other partition, the subscription column family does not contain such an entry.
Hypothesis: A `MESSAGE_SUBSCRIPTION/CORRELATE` or  `MESSAGE_SUBSCRIPTION/DELETE`command was lost, leaking the subscription entry. `ProcessMessageSubscriptionCorrelateProcessor` handles non interrupting subscriptions by removing them from the state immediately and then sending `MESSAGE_SUBSCRIPTION/CORRELATE` just once. If that command is lost, nothing will resend it. The other partition keeps the partition in the state, tries to correlate again and again but only gets rejections because the receiving partition doesn't know about the subscription anymore.
 
## Related Notes

- It probably doesn't make sense to reject a `MESSAGE_SUBSCRIPTION/REJECT` command, can't we just clean up the subscription when the receiving partition rejected it?
- Why write `COMMAND_REJECTION`s for internal commands not sent by a user anyway? Especially since the rejection is already described with a `MESSAGE_SUBSCRIPTION/REJECT` command.
- To fully fix this, we need to find a way to recover from this broken state, not just prevent it from happening again.

### PR DESCRIPTION
**Pull request by lenaschoenburg**
This fixes two issues with rejecting message correlation.

First, there appears to be a long-standing bug where `MESSAGE_SUBSCRIPTION/REJECT` was written on the process instance partition, not the message partition. I couldn't find a state in the git history where this has not been the case but best I can tell, `MESSAGE_SUBSCRIPTION` records are always supposed to be written on the message partition.

:warning: While it makes perfect sense to me that this was a bug and my fix is correct, I'm also very surprised that this was not caught before and that message correlation appeared to work. A double check by someone more familiar with message correlation is required.

Second is the bug reported in #13814 where a missed acknowledgement (via `MESSAGE_SUBSCRIPTION/CORRELATE`) would result in an endless loop of correlation attempts and rejections. To fix this, a `MESSAGE_SUBSCRIPTION/REJECT` is now treated as a definitive signal that the subscription is not supposed to exist anymore. As a result, the message partition now removes the subscription when correlation is rejected and stops trying to correlate.

:warning: While I've discussed the bug fix for #13814 with @saig0, I'm still not entirely sure if the new behavior for `MESSAGE_SUBSCRIPTION/REJECT` is always safe. The risk I see is that we might accidentally remove a subscription that was not supposed to be removed. I'm not familiar enough with BPMN and the engine implementation to figure that out.

:sos: I've tried adding a test to verify the behavior when acknowledgments are lost (the root cause for #13814). Overall I'm not happy with this test, mainly because I'm not sure if it is verifying enough and also because it is very slow. Some help to improve this test would be greatly appreciated <3.

Closes #14814

### CODE REVIEW

**Comment by saig0** [comment]
> appears to be a long-standing bug where `MESSAGE_SUBSCRIPTION/REJECT` was written on the process instance partition, not the message partition

You're right. :see_no_eval:

> I'm also very surprised that this was not caught before and that message correlation appeared to work. 

`MESSAGE_SUBSCRIPTION/REJECT` is not the common case. The REJECT command is only written if we close the message subscription (i.e. the message catch event is left) and correlate the message concurrently. In most cases, the message is either correlated or the subscription is closed.

> I'm still not entirely sure if the new behavior for MESSAGE_SUBSCRIPTION/REJECT is always safe. The risk I see is that we might accidentally remove a subscription that was not supposed to be removed.

I'm confident that the new behavior is safe and we can remove the subscription on a REJECT command. When we send the REJECT command, it should not be possible to correlate a message for this subscription. I don't see any other case at the moment.

**Comment by remcowesterhoud** [comment]
@oleschoenburg what is the status of this PR? It's still in draft but you also requested our review. Do you want us to review it before it's ready?

**Comment by oleschoenburg** [comment]
@remcowesterhoud thanks for the ping. PR is ready for review, I just forgot to remove the draft status.

File: engine/src/test/java/io/camunda/zeebe/engine/util/EngineRule.java [hunk 1]
```diff
@@ -405,6 +407,14 @@ public EngineRule maxCommandsInBatch(final int maxCommandsInBatch) {
     return this;
   }
 
+  public void interceptInterPartitionCommands(final CommandInterceptor interceptor) {
```
Comment (remcowesterhoud) [inline engine/src/test/java/io/camunda/zeebe/engine/util/EngineRule.java:410]: 🤩 Nice!

File: engine/src/test/java/io/camunda/zeebe/engine/processing/message/MessageCorrelationReliabilityTest.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        .isOne();
+  }
+
+  private int getPartitionId(final String correlationKey) {
+    final List<Integer> partitionIds = engine.getPartitionIds();
+    return SubscriptionUtil.getSubscriptionPartitionId(
+        BufferUtil.wrapString(correlationKey), partitionIds.size());
+  }
+
+  private ProcessInstanceWaitingForMessage waitingProcessInstance(
```
Comment (remcowesterhoud) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/message/MessageCorrelationReliabilityTest.java:150]: 🔧 This method is unused

**Review by remcowesterhoud — APPROVED** [review]
@oleschoenburg Thanks for the PR and sorry it took so long to review 🙇 

I have a small comment about the test case.
I've also taken the liberty to speed it up a bit. Please have a look at[ my latest commit](https://github.com/camunda/zeebe/pull/14820/commits/4a86f03a08e08a5a631adfe9867db869f07f060b) to see how 👀

**Comment by oleschoenburg** [comment]
Awesome, thanks @remcowesterhoud <3

bors merge

**Pull request merged**